app/models/test_track/session.rb

```diff
@@ -16,6 +16,27 @@ def manage
     notify_unsynced_assignments! if sync_assignments?
   end
 
+  def visitor_dsl_for(identity)
+    if has_matching_identity?(identity)
+      visitor_dsl
+    else
+      TestTrack::VisitorDSL.new(visitors_by_identity[identity])
+    end
+  end
+
+  def visitors_by_identity
+    @visitors_by_identity ||= Hash.new do |visitors_by_identity, identity|
+      remote_visitor = TestTrack::Remote::Visitor.from_identifier(
+        identity.test_track_identifier_type,
+        identity.test_track_identifier_value
+      )
+      visitors_by_identity[identity] = TestTrack::Visitor.new(
+        id: remote_visitor.id,
+        assignments: remote_visitor.assignments
+      )
+    end
```

Comment:
So I think this is clever but in a not awful way - using a custom Hash initialization to auto-add to the collection whenever somebody requests a visitor for an identity that we've never seen before.

Maybe this should be private so nobody does anything naughty to the hash's state from outside the model.

Reply:
my vote is to extract this Hash into a class that has an easier-to-understand interface than the self-populating hash thing we're doing here. i'm fine with waiting to do that until we get rid of the `SessionIdentityCollection` so that you can avoid any confusion about what that class is vs. this new one i'm proposing.